ath79: preliminary support for TP-Link WDR3600 / WDR4300 (AR9344)

juppin opened the pull request:
### working
 - leds
 - buttons
 - lan / wan
 - usb ports 1 + 2
 - wifi 5g
 - sysupgrade
 - ...

### not working
 - wifi 2g

### bootlog with two usb sticks attached
https://pastebin.com/XgeRN1pD

cpu is overclocked to 700 mhz, so the cpu speed in bootlog is real and not a error.

usb shows "not running at top speed" but speed is almost the same as on ar71xx

### usb benchmark
```
time dd if=/dev/zero of=/dev/sda1 bs=10M count=10 conv=fsync
10+0 records in
10+0 records out
real    0m 38.76s
user    0m 0.02s
sys     0m 1.74s
100M / 38.76 s = 2.58 M/s
```

### iperf nat benchmark (tcp) wan -> lan

with flow offloading:
```
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-30.0 sec  3.24 GBytes   929 Mbits/sec
```

without flow offloading:
```
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-30.0 sec  2.13 GBytes   611 Mbits/sec
```

### iperf nat benchmark (tcp) wan -> wifi 5g

with flow offloading (~30% idle):
```
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-30.0 sec   680 MBytes   190 Mbits/sec
```

without flow offloading (~0% idle):
```
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-30.0 sec   539 MBytes   151 Mbits/sec
```

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

jneuhauser committed:
ath79: preliminary support for TP-Link WDR3600 / WDR4300 (AR9344)

working:
 - leds
 - buttons
 - lan / wan
 - usb (hub port 1 + 2)
 - wifi 5g
 - sysupgrade
 - ...

not working:
 - wifi 2g

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

juppin commented:
If anyone want to test a precompiled image... 
[openwrt-ath79-generic-tl-wdr.zip](https://github.com/openwrt/openwrt/files/2017807/openwrt-ath79-generic-tl-wdr.zip)
please left feedback if you are testing my prebuild images...

juppin commented:
> Could you perhaps point me to a summary of the issues with 2Ghz WiFi?

Take a look at the commit history of the ath79 target and read the first commit of blogic...
https://github.com/openwrt/openwrt/commits/master/target/linux/ath79

> Is there any sort of outline for porting an existing router from ar71xx to ath79?

Reading mach files and source code of ar71xx target... Reading dt docs, source code and dts of the new ath79 target. And "grep -r KEYWORD" to search in the kernel source and docs...
With this pr, it should be very easy to get your DIR-835-A1 (Very similar to WDR3600/WDR4300) ported to the new ath79 target. You should be able to get all except soc wireless working...

> Also, do you know if it's possible to test these kernels using kexec, to minimize chances of borking one's device?

Never tried that before with openwrt... At least you have to use the initram image and cut off the 64 byte uImage header (512 byte for tplink header)...

For my tests of hw except flash, i use tftpboot from within u-boot... For testing flash you have to flash it.
Before you start to port your device, ensure you could use recovery over tftp or http and do a backup of all your mtd partitions.
Your DIR-835-A1 should have a http recovery implemented in u-boot.

lucize commented:
@guidosarducci You could try the *-initramfs-kernel.bin from uboot (with proper tftp server and filename), maybe it should run using `bootm`
1043nd-v1 works from memory just by using
```
tft
bootm
```

juppin commented:
@blogic Is there a reason while "ath79_set_pll(ag)" is for now only called for "qca,ar7100-eth" and "qca,ar7242-eth"?

blogic commented:
@juppin no, needs to be added, same as set_speed and possible set_gmac/mii inside the adjust_link callback

blogic commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -21,6 +21,12 @@ case "$board" in
+"tplink,tl-wdr3600"|\
+"tplink,tl-wdr4300")
+	ucidef_set_led_usbdev "usb1" "USB1" "tp-link:green:usb1" "1-1.1"
+	ucidef_set_led_usbdev "usb2" "USB2" "tp-link:green:usb2" "1-1.2"
```
put these into dts directly please.

juppin replied:
Did you have a hint for me how to achieve this inside dts for the two ports ?

lucize replied:
I had an email with this:

I've a request. Please don't use ucidef_set_led_usbdev but put info
about USB ports & LEDs in DT instead. For a reference you may check:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=69d22c70ac9ad66be671ad2517ad5ee41058255f
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=0b1f11002a70abf958aacbf4c18858443f504986

juppin replied:
Great, does work... Thanks!

blogic commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -21,6 +21,12 @@ case "$board" in
+"tplink,tl-wdr4300")
+	ucidef_set_led_usbdev "usb1" "USB1" "tp-link:green:usb1" "1-1.1"
+	ucidef_set_led_usbdev "usb2" "USB2" "tp-link:green:usb2" "1-1.2"
+	ucidef_set_led_wlan "wlan2g" "WLAN2G" "tp-link:green:wlan2g" "phy0tpt"
```
this should also be possible to encode inside OF. if not let me know and we'll add support

juppin replied:
With "trigger-sources" as for usb it is not possible.
For now i set the wireless leds with "$(get_dt_led wifi)" and an alias for the led.
Would be great if it will also be possible with "trigger-sources".

blogic commented:
looking good, just did a mini review. wondering if we can put the ath10k eep into dts aswell

juppin commented:
If i look in the source code for ath9k, it isn´t possible to get the eeprom from mtd with a defined property in dt.
All except eeprom and switch config can now set through dt.

lucize commented:
aren't this [patch series](https://patchwork.kernel.org/patch/10306449/) meant for that ?

juppin commented:
@lucize it sounds like that series is exactly for that propose

@blogic if you have time, would you take a look on eeprom and wireless led bindings?

blogic commented:
@juppin sure, we can use the existing script hooks for now

juppin commented:
I have to correct me, setting property `linux,default-trigger = "phy0tpt";` inside led node does use it as trigger for wireless activity...

juppin commented:
I did a complete review of my complete commit and corrected some small errors...
- wifi btn as switch
- removed unused labels

PR is ready for a review.

blogic commented:
@juppin nice work, merged, Thanks !

juppin commented:
@blogic great. I have to say thanks to all the great openwrt devs :-)

guidosarducci commented:
@juppin @lucize  Thanks very much for the feedback regarding DIR-835-A1. I'd already been looking through some of the sources you mentioned, but hoped for something overarching to "connect-the-dots". I'll continue doing that, and also think about a good test setup or 2nd unit.
